### PR TITLE
Make variable expansion work for environment variables in k8s

### PR DESCRIPTION
**Pull request by metlos**
### What does this PR do?
K8s does the expansion only if it already knows about the variable being
expanded.

This means we have to sort the environment variable list prior to sending
it to k8s in such a way that vars that reference others always follow the
referenced ones.

This is a reimplementation of #12610 that uses an appropriate algorithm to produce a consistent sorting of the env vars.

### What issues does this PR fix or reference?

#12577

#### Release Notes
* [ ] TODO

#### Docs PR
N/A

### CODE REVIEW

**Comment by metlos** [comment]
ci-test

**Comment by metlos** [comment]
ci-build

**Comment by metlos** [comment]
ci-test

**Review by skabashnyuk — APPROVED** [review]

**Comment by metlos** [comment]
ci-test

**Comment by metlos** [comment]
ci-test

**Comment by metlos** [comment]
@sleshchenko @garagatyi could you please take a look and review this?

File: core/commons/che-core-commons-lang/src/test/java/org/eclipse/che/commons/lang/TopologicalSortTest.java
Comment (sleshchenko) [inline core/commons/che-core-commons-lang/src/test/java/org/eclipse/che/commons/lang/TopologicalSortTest.java:123]: It needs some time to understand that `name` is one character string, and value is multi-characters strings that actually contains dependency. Would not it be simpler if it would look like the following:
```java
  private static final class Var {

    private final Character name;
    private final Set<Character> dependencies;
    ...
}

```
WDYT?

File: core/commons/che-core-commons-lang/src/test/java/org/eclipse/che/commons/lang/TopologicalSortTest.java
Comment (sleshchenko) [inline core/commons/che-core-commons-lang/src/test/java/org/eclipse/che/commons/lang/TopologicalSortTest.java]: What is the difference between `sort dependents after dependencies` cases?
It would be better to see descriptions like
`// sort dependents after dependencies. case 1: 2 independent + 1 dependent on one`
`// sort dependents after dependencies. case 1: 2 independent + 1 dependent on two others`
etc.

File: core/commons/che-core-commons-lang/src/test/java/org/eclipse/che/commons/lang/TopologicalSortTest.java
Comment (sleshchenko) [inline core/commons/che-core-commons-lang/src/test/java/org/eclipse/che/commons/lang/TopologicalSortTest.java:39]: Here is a valid test case
```java
      // removes duplicates
      new Object[] {
          vars(var("a", "."), var("a", "."), var("b", ".")),
          vars(var("a", "."), var("b", "."))
      }
```
it is expected? if yes - please add note in java doc

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java
Comment (sleshchenko) [inline core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java]: You can avoid duplicating this code 
```java
removePredecessorMapping(predecessors, successors, curr);
results.add(curr);
```
Like the following
```java
...
        TreeSet<ID> cycle;
...
        // If we ever find a graph that doesn't have any independent node, yet we fail to find a
        // cycle in it, the universe must be broken.
        if (cycle.isEmpty()) {
          throw new IllegalStateException(
              String.format(
                  "Failed to find a cycle in a graph that doesn't seem to  have any independent"
                      + " node. This should never happen. Please file a bug. Current state of the"
                      + " sorting is: predecessors=%s, successors=%s, results=%s",
                  predecessors.toString(), successors.toString(), results.toString()));
        }

        // remove the cycle from the predecessors and update the predecessor map so that dependent
        // nodes that are not part of the cycle can be sorted properly
        // cycle elements are inserted in result in order they were found but not original order
        for(ID id : cycle) {
          removePredecessorMapping(predecessors, successors, id);
          results.add(id);
        }
      }
    }
  }

  private void removePredecessorMapping(
      Map<ID, Set<ID>> predecessors, Map<ID, Set<ID>> successors, ID node) {
    for (ID succ : successors.getOrDefault(node, emptySet())) {
      predecessors.getOrDefault(succ, emptySet()).remove(node);
    }
    predecessors.remove(node);
  }

  private TreeSet<ID> findCycle(ID node, Map<ID, Set<ID>> predecessors) {
    // bail out quickly if there are no preds - should be fairly common occurrence hopefully
    Set<ID> preds = predecessors.get(node);
    if (preds == null || preds.isEmpty()) {
      return new TreeSet<>();
    }

    TreeSet<ID> ret = new TreeSet<>();
    // start node is first item of result
    ret.add(node);
    List<ID> todo = new ArrayList<>(preds);

    Set<ID> visited = new HashSet<>();

    while (!todo.isEmpty()) {
      ID n = todo.remove(0);

      if (visited.contains(n)) {
        continue;
      }

      visited.add(n);

      todo.addAll(predecessors.getOrDefault(n, emptySet()));
      ret.add(n);

      if (n.equals(node)) {
        // we found the cycle to our original node
        return ret;
      }
    }

    return new TreeSet<>();
  }
...
```
WDYT?

**Review by sleshchenko — APPROVED** [review]
Good job :+1: 

Really great javadocs and comments that help understand the implementation

File: core/commons/che-core-commons-lang/src/test/java/org/eclipse/che/commons/lang/TopologicalSortTest.java
Comment (sleshchenko) [inline core/commons/che-core-commons-lang/src/test/java/org/eclipse/che/commons/lang/TopologicalSortTest.java:67]: Just one more small proposal for cases names
```java
// item depends on itself
// cycle with 2 items
// cycle with 3 items + 1 independent item
```
The last case is quite tricky not sure if I can easily clear formulate description.

**Comment by metlos** [comment]
ci-test

File: core/commons/che-core-commons-lang/src/test/java/org/eclipse/che/commons/lang/TopologicalSortTest.java
Comment (metlos) [reply core/commons/che-core-commons-lang/src/test/java/org/eclipse/che/commons/lang/TopologicalSortTest.java:67]: explanations added, hopefully they're intelligible :)

**Comment by metlos** [comment]
I squashed and rebased on top of latest master to clean up the PR a little

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvVarsConverter.java
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvVarsConverter.java]:
```suggestion
        // The environment variable expansion only works if a variable that is referenced
```

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvVarsConverterTest.java
Comment (garagatyi) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvVarsConverterTest.java]: There is `assertEqualsNoOrder` for such cases. Unfortunately it check only arrays, but you can convert lists to arrays with `list.toArray()`. This would not change much, but the test would be a little bit more concise.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/EnvVars.java
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/EnvVars.java:23]: There are also variable notation with curly brackets, e.g. `export MY_VAR=${ANOTHER_VAR}`
As far as I understand this class won't support it. Is it deliberately made decision to skip support of it?

Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/EnvVars.java:23]: It seems this pattern supports `$(some)` which is rather command substitution. See http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_03
Can you elaborate why we don't support more common cases of expansion like `$MY_VAR` or `${MY_VAR}`?

Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/EnvVars.java:23]: And since it is command substitution it should not even work like that for regular shells. So, it is completely custom feature that just looks like what shell do.

Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/EnvVars.java:23]: Should we implement this feature with a custom notation or use something widely known like shell expansion? 
@metlos @sleshchenko @skabashnyuk @l0rd WDYT?

Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/EnvVars.java:23]: This is specifically for variable expansion in kubernetes, which only allows the `$(...)` syntax. See https://github.com/kubernetes/kubernetes/blob/master/third_party/forked/golang/expansion/expand.go

Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/EnvVars.java:23]: Or the docs: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#envvar-v1-core

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvVarsConverterTest.java
Comment (metlos) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvVarsConverterTest.java]: `assertEqualsNoOrder` seems to not be available in TestNG assert. I made a change to the test code anyway so that the purpose is hopefully slightly more clear.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/EnvVars.java
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/EnvVars.java:23]: Since we assume k8s-based infrastructure, I think we should stick to the conventions available there.

Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/EnvVars.java:23]: Now it is clear. Thanks for the pointer. BTW here is the corresponding user docs https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/#using-environment-variables-inside-of-your-config
Could you add a link to a doc to the source code to make it less confusing to those who don't know that k8s uses special notation?

Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/EnvVars.java:23]: Good idea

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvVarsConverterTest.java
Comment (garagatyi) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvVarsConverterTest.java]: Now it is cleaner, agree. But arrays assertion should be in test ng. See http://static.javadoc.io/org.testng/testng/6.0.1/org/testng/Assert.html#assertEqualsNoOrder(java.lang.Object[],%20java.lang.Object[])
Maybe it will be useful next time

Comment (metlos) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvVarsConverterTest.java]: Huh, I'm blind :wink: Thanks for pointing it me to the method, I didn't know about it before. In any case, it would not be acceptable here anyway, because the test is specifically for the order of the elements, checking that the env vars are topologically sorted.

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java
Comment (garagatyi) [inline core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java]: FYI If the intention to precisely predict the size of the map than just size might not be a good predictor. This is the reason why Google Guava has methods for creation of maps with predicted size. But this should not be a problem with non-highload projects and could be treated as premature optimization.

Comment (metlos) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java]: In here we know the size **precisely**, though. Notice that for example for the successors I use the default ctor `HashSet<>()` because I have no way of predicting the target size.

IMHO, trying to "guess" an initial/needed size of the collections is not a premature optimization but just a good practice. In https://github.com/rhq-project/rhq we were able to reduce the memory consumption by 30% (!!!) by just thinking about and limiting the initial capacity of the collections involved.

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java
Comment (garagatyi) [inline core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java]: I can misunderstand but isn't it gets removed in `removeFirstIndependent`? And as far as I know LinkedHashMap throws ConcurrentModificationException if iterator is used after map was changed aside of the iterator.
But if this is intended behavior for different cases then it would be great to leave a comment about that in the code - that remove won't be performed in the iterator case and exceptional case would not occur.

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java
Comment (garagatyi) [inline core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java:141]: This comment says what needs to be done and it is done one line below. It could be better to explain why it needs to be done, but not what needs to be done
```suggestion
        // Independent node doesn't need processing
```

**Comment by metlos** [comment]
Squashed commits together before merging.

**Review by garagatyi — COMMENTED** [review]
It seems I failed to understand the algorithm to the end. Can't approve or requests changes because of that. I left some comments, but rely on other reviewers reviews on whether to merge right away.

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java
Comment (garagatyi) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java]: In my opinion, usage of a precise size of another map is not reliable to set capacity of the map. The reason that capacity != size. With the default load factor 0.75 capacity is 25% bigger. See https://stackoverflow.com/questions/30220820/difference-between-new-hashmapint-and-guava-maps-newhashmapwithexpectedsizein

Sorry if I'm missing some important point why this work otherwise in your case. And would be happy to understand this maps sizing issue better.

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvVarsConverterTest.java
Comment (garagatyi) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvVarsConverterTest.java]: Oh, yeah, you are right. I missed that we need to check the order

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java
Comment (metlos) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java]: Ok, seems like you got me here :)

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java
Comment (metlos) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java]: Excellent observation! The CME wouldn't occur, because the exact same entry is removed by the iterator in `removeFirstIndependent` prior to calling this method, so the map is never modified by this call. It is just wasteful to call it. I've modified the impl to take this into account.

**Comment by metlos** [comment]
ci-build

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java
Comment (metlos) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java]: I actually thought that the behavior was the same for `*List(int capacity)` as well as `*Set` and `*Map` cases. Thanks for pointing out the difference to me.

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java
Comment (metlos) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java]: This is not bit manipulation since the variable is a boolean, but I have changed it. I agree your suggestion is easier to read.

Comment (garagatyi) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java]: Oh, I didn't know that this is specified in the language spec https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.22.2
Thanks for sharing!
But again, this might not be very well known stuff. In a case, when XOR is needed I would go with ^.

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java
Comment (metlos) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java:141]: IMHO, both my comment and your suggestion add no essential information :wink: The reason I tend to add comments like this is to highlight the difference between the simple/fast and complicated/slow case.

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java
Comment (garagatyi) [reply core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java]: I didn't mean to get you. This is a tricky part of java that we discussed a lot back when I was in the platform team, so now we can share this knowledge.

**Pull request merged**
